openpgp/packet/public_key.go

```diff
@@ -35,10 +35,11 @@ type kdfAlgorithm byte
 
 // PublicKey represents an OpenPGP public key. See RFC 4880, section 5.5.2.
 type PublicKey struct {
+	version      int
```

Comment:
This was my initial approach, however:
- I think having internal/unexported `version` is better API design, since the version gets set internally, and allows to set a default in only one place the `Version()` function. Moreover, this prevents version downgrades on keys/signatures that could lead to i.e. checking key Ids instead of fingerprints and losing properties of `v5` keys. Instead, the user can't modify the version (only input keys/signatures ofc), and if the code runs into an odd version (i.e., 0 - not set), it just defaults to whatever (currently, 4) and tries with that, or errors.
- There are some changes to make in the code. For example. `NewEntity` it needs to set e.g. version 4 (so that it is not 0). This involves setting the version in functions like (RSA) `GenerateKeyWithPrimes` or adding some logic before (`newSigner`).

It can be done but I largely prefer this design.

Reply:
Yeah, but I think having `version` as `0` and defaulting to `4` is kind of hacky; I would rather have `NewEntity` set the version explicitly, anyway. And then, instead of setting it to 4 and calling `UpgradeToV5()`, it could just directly set it to the version that it wants.

Code that has access to the packet object can already do lots of strange things if it wants, so I don't think that the fact that it can modify the packet version is a huge issue. There are lots of other ways it could break the key / break security.

Reply:
I'll export Version then, although don't like contributing further to strange behavior. Also, I'll need to modify every test initializing a public key or signature in order to include the version parameter 
(this is what I meant by the `0` version; an application setting e.g. 
```
pk := &PublicKey{<values>}
```
if it does not set `Version` then we should assume there's a default instead of failing with `public key version 0` somewhere).

Reply:
(also, we need to modify all of `New{RSA, DSA, ElGamal, ECDSA, ECDH, EdDSA}PublicKey`, since they're part of the API they need to output keys with the correct `PublicKey.Version`)

Reply:
But also when they do `&PublicKey`, they would have to still call `setFingerprintAndKeyId`, right? So creating the packet object manually is a bit useless, and everyone should already be using `NewRSAPublicKey`. And then for `New{RSA, DSA, ElGamal, ECDSA, ECDH, EdDSA}PublicKey`; yeah, it is a bit unfortunate that they don't have a `config` parameter. Maybe for now, we could just set `Version: 4` in all of them, and then upgrade to v5 in `NewEntity` after all, indeed. Even though I think it's a bit ugly, but people that call `NewRSAPublicKey` might expect a v4 key, anyway.

Reply:
Yes, you cannot call `setFingerprintAndKeyId` from the outside, but you can always set the fingerprint and key ID directly. If some user is doing so, this change is breaking (they'll need to put `Version: 4`).

Reply:
Yeah, but.. I think it's fair to say that the API as is expects you to call `NewRSAPublicKey` etc, and that someone who isn't doing so is just using the API wrong :sweat_smile:

Reply:
Right, still an API breaking change to consider (maybe someone who reads the values from somewhere else (say, not in armored format) and doesn't need to generate a new key?); the new `Version` field becomes compulsory.